### DISCUSSION TITLE
"Visited Links" can't be removed in the running App

### DISCUSSION THREAD

**Discussion by hhu-mahmoud** [cefsharp/CefSharp#4361] [General]
### Bug Report

- **What version of the product are you using?**
     Version: 108.4.130

- **What architecture x86 or x64?**
    x64
	
- **What version of .Net?**
    .Net 7.0
    
- **On what operating system?**
    Win10

- **Are you using `WinForms`, `WPF` or `OffScreen`?**
    WPF
    
- **What steps will reproduce the problem?**
   
Remove "Cache" folder, while the system is Running and after Cef Shutdown, Browser Close and Disapose.
If I tried to remove the "Cache" folder while the system is running and after I shutdown the Cef and Close all open browsers and disapose them, then I get a exception, that only the file "Visited Links" is still open by another process. I could delete the file only if the App is totally/complete closed.
   
- **What is the expected output? What do you see instead?**

I need to delete the "Cache" folder and its content also "Visited Links" by closing the App. Everything would be delete unless the file "Visited Links". Do you have an solution for this problem or any ideas?

**Comment by amaitland** [comment]
Converted to discussion. `Visited Links` is entirely implemented at the `Chromium` layer, it's not something that `CefSharp` has direct control over.

There's a related discussion at https://www.magpcss.org/ceforum/viewtopic.php?f=6&t=18686

> I need to delete the "Cache" folder and its content

What do you need to delete the cache for? You can use an `InMemory` cache if you don't want to persist data to disk.

You can also clear the `browser cache` and `cookies` at runtime via `CDP` if required.

**Comment by hhu-mahmoud** [comment]
@amaitland 

The user can the deactivate and activate the cache in the program. If the user clicks on cache deactivate, then should the stop tracking the data in cache and by closing the app should the cache folder removed.

- By closing the program:

`if (Cef.IsInitialized)
        {
            Cef.ShutdownWithoutChecks();
        }`
        
and also i check in browser object is exist, then i disapose it.
        
I use also this command to remove the cache path, if the user has deactivated the cache:
        
`RequestContext.Configure().WithCachePath("").Create();`
        
after that i delete the cache folder from AppData/Romaing:
        
```
// TODO: Visited links can not be removed while the system running!!
 if (!DeactivateBrowserCaching || !Directory.Exists(_cacheDir))
    {
      return;
     }
// delete the cache folder if exists after the browser shutdown
Directory.Delete(_cacheDir, true);
```

I don't need any cache folder in AppData/Romaing if the user don't want to use the cache!!

In this link, the solution is to build a helper service to clean everything after the app totally closed. I don't want to build a cleanup service for just this file "Visited Links" and i don't understand, why it still open at all and how may i close it?!

**Reply by amaitland** [reply]
> if (Cef.IsInitialized) { Cef.ShutdownWithoutChecks(); }

`Cef.Shutdown` should be used instead. If you changed to using `ShutdownWithoutChecks` because you were getting an exception then you are calling `Shutdown` on the incorrect thread and that's possibly part of your problem. 

You can also try enabling `Cef.EnableWaitForBrowsersToClose()` and calling `Cef.WaitForBrowsersToClose();` before `Cef.Shutdown()` is called.

See https://github.com/cefsharp/CefSharp/issues/3047#issuecomment-623225981

> and i don't understand, why it still open at all and how may i close it?!

You would need to debug `Chromium` to determine the reason it's still open. As I've already said `CefSharp` has no direct control over this  behaviour. 

> In this link, the solution is to build a helper service to clean everything after the app totally closed

You don't need anything fancy, just spawn your app with a command line arg that tells it to cleanup, wait for your main application to exit then delete the folder. You don't need a **service**.

**Comment by hhu-mahmoud** [comment]
Problem solved by deleting the cache folder after closing the app.
